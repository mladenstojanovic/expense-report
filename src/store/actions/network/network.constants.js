export const GET_AUTH_TOKEN_START = '[NETWORK] Get Auth Token Start';
export const GET_AUTH_TOKEN_SUCCESS = '[NETWORK] Get Auth Token Success';
export const GET_AUTH_TOKEN_ERROR = '[NETWORK] Get Auth Token Error';
export const CREATE_USER_START = '[NETWORK] Create User Start';
export const CREATE_USER_SUCCESS = '[NETWORK] Create User Success';
export const CREATE_USER_ERROR = '[NETWORK] Create User Error';
export const ADD_BANK_CONNECTION_START = '[NETWORK] Add Bank Connection Start';
export const ADD_BANK_CONNECTION_SUCCESS =
  '[NETWORK] Add Bank Connection Success';
export const ADD_BANK_CONNECTION_ERROR = '[NETWORK] Add Bank Connection Error';
export const GET_TRANSACTIONS_START = '[NETWORK] Get Transactions Start';
export const GET_TRANSACTIONS_SUCCESS = '[NETWORK] Get Transactions Success';
export const GET_TRANSACTIONS_ERROR = '[NETWORK] Get Transactions Error';
export const NETWORK_OPERATIONS_START = '[NETWORK] Network Operations Start';
export const CONNECTION_JOB_START = '[NETWORK] Connection Job Start';
export const CONNECTION_JOB_SUCCESS = '[NETWORK] Connection Job Success';
export const CONNECTION_JOB_ERROR = '[NETWORK] Connection Job Error';

export const IN_PROGRESS = 'in progress';
export const SUCCESS = 'success';
export const ERROR = 'error';

export const RETRIEVE_TRANSACTIONS = 'retrieve-transactions';
