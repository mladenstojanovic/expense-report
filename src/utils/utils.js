import React from 'react';
import { render } from '@testing-library/react';
import { createStore, combineReducers } from 'redux';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import networkReducer from '../store/reducers/network.reducer';
import theme from '../styles/theme';

export const sleep = milliseconds => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const sortTransactionData = data => {
  const sortedData = {};
  for (let index = 0; index < data.length; index++) {
    if (!data[index].subClass) {
      continue;
    }
    if (!sortedData[data[index].subClass.code]) {
      sortedData[data[index].subClass.code] = {
        total: 0,
        numOfTransactions: 0,
        title: data[index].subClass.title
      };
    }
    sortedData[data[index].subClass.code].total += Math.abs(
      parseInt(data[index].amount)
    );
    sortedData[data[index].subClass.code].numOfTransactions++;
  }
  return sortedData;
};

export const isEmailValid = email => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);

export const isPhoneNumberValid = phone =>
  /^[+]?[(]?[0-9]{3}[)]?[-\s.]?[0-9]{3}[-\s.]?[0-9]{4,6}$/.test(phone);

export const renderWithReduxAndStyles = (
  ui,
  {
    initialState,
    store = createStore(
      combineReducers({ network: networkReducer }),
      initialState
    )
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Provider>
  ),
  store
});
