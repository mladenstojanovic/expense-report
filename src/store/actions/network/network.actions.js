import {
  GET_AUTH_TOKEN_START,
  CREATE_USER_START,
  ADD_BANK_CONNECTION_START,
  GET_TRANSACTIONS_START,
  GET_AUTH_TOKEN_SUCCESS,
  CREATE_USER_SUCCESS,
  ADD_BANK_CONNECTION_SUCCESS,
  GET_TRANSACTIONS_SUCCESS,
  NETWORK_OPERATIONS_START,
  CONNECTION_JOB_START,
  CONNECTION_JOB_SUCCESS,
  GET_AUTH_TOKEN_ERROR,
  CREATE_USER_ERROR,
  ADD_BANK_CONNECTION_ERROR,
  CONNECTION_JOB_ERROR,
  GET_TRANSACTIONS_ERROR,
  SUCCESS,
  RETRIEVE_TRANSACTIONS
} from './network.constants';
import { sleep, sortTransactionData } from '../../../utils/utils';

export const getAuthTokenStart = () => ({
  type: GET_AUTH_TOKEN_START
});

export const getAuthTokenSuccess = () => ({
  type: GET_AUTH_TOKEN_SUCCESS
});

export const getAuthTokenError = () => ({
  type: GET_AUTH_TOKEN_ERROR
});

export const createUserStart = () => ({
  type: CREATE_USER_START
});

export const createUserSuccess = () => ({
  type: CREATE_USER_SUCCESS
});

export const createUserError = () => ({
  type: CREATE_USER_ERROR
});

export const addBankConnectionStart = () => ({
  type: ADD_BANK_CONNECTION_START
});

export const addBankConnectionSuccess = () => ({
  type: ADD_BANK_CONNECTION_SUCCESS
});

export const addBankConnectionError = () => ({
  type: ADD_BANK_CONNECTION_ERROR
});

export const connectionJobStart = () => ({
  type: CONNECTION_JOB_START
});

export const connectionJobSuccess = () => ({
  type: CONNECTION_JOB_SUCCESS
});

export const connectionJobError = () => ({
  type: CONNECTION_JOB_ERROR
});

export const getTransactionsStart = () => ({
  type: GET_TRANSACTIONS_START
});

export const getTransactionsSuccess = payload => ({
  type: GET_TRANSACTIONS_SUCCESS,
  payload
});

export const getTransactionsError = () => ({
  type: GET_TRANSACTIONS_ERROR
});

export const networkOperationsStart = () => ({
  type: NETWORK_OPERATIONS_START
});

let userId;
let token;
let parsedAddConnectionResponse;
let jobFinished = false;

export const getAuthToken = async dispatch => {
  try {
    dispatch(networkOperationsStart());
    dispatch(getAuthTokenStart());

    const authResponse = await fetch(process.env.REACT_APP_TOKEN_ENDPOINT, {
      headers: {
        Authorization: `Basic ${process.env.REACT_APP_API_KEY}`
      }
    });
    const authResponseParsed = await authResponse.json();

    token = authResponseParsed.access_token;
    dispatch(getAuthTokenSuccess());
  } catch {
    dispatch(getAuthTokenError());
  }
};

export const createUser = async (dispatch, userData) => {
  try {
    dispatch(createUserStart());

    const createUserResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users`,
      {
        method: 'POST',
        body: JSON.stringify(userData),
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const createUserResponseJson = await createUserResponse.json();

    userId = createUserResponseJson.id;
    dispatch(createUserSuccess());
  } catch {
    dispatch(createUserError());
  }
};

export const addConnection = async dispatch => {
  try {
    const testUser = {
      loginId: process.env.REACT_APP_LOGIN_ID,
      password: process.env.REACT_APP_LOGIN_PWD,
      institution: {
        id: process.env.REACT_APP_LOGIN_INSTITUTION_ID
      }
    };

    dispatch(addBankConnectionStart());

    const addConnectionResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/connections`,
      {
        method: 'POST',
        body: JSON.stringify(testUser),
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      }
    );
    parsedAddConnectionResponse = await addConnectionResponse.json();

    // mozda neki error handling?

    dispatch(addBankConnectionSuccess());
  } catch {
    dispatch(addBankConnectionError());
  }
};

export const connectionJob = async dispatch => {
  try {
    dispatch(connectionJobStart());
    //@TODO: config?
    let retries = 20;
    do {
      //@TODO: config?
      await sleep(10000);

      const job = await fetch(`${parsedAddConnectionResponse.links.self}`, {
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      });
      const parsedJob = await job.json();
      const retrieveTransactionJob = parsedJob.steps.find(
        job => job.title === RETRIEVE_TRANSACTIONS
      );

      if (retrieveTransactionJob.status === SUCCESS) {
        jobFinished = true;
      } else {
        retries--;
      }
    } while (!jobFinished && retries > 0);
    if (jobFinished) {
      dispatch(connectionJobSuccess());
    } else {
      dispatch(connectionJobError());
    }
  } catch {
    dispatch(connectionJobError());
  }
};

export const getTransactions = async dispatch => {
  try {
    dispatch(getTransactionsStart());
    const getTransactionsResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/transactions`,
      {
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const parsedGetTransactionsResponse = await getTransactionsResponse.json();

    dispatch(
      getTransactionsSuccess(
        sortTransactionData(parsedGetTransactionsResponse.data)
      )
    );
  } catch {
    dispatch(getTransactionsError());
  }
};

export const submitUser = userData => {
  return async dispatch => {
    await getAuthToken(dispatch);
    await createUser(dispatch, userData);
    await addConnection(dispatch);
    await connectionJob(dispatch);
    if (jobFinished) {
      await getTransactions(dispatch);
    }
  };
};
