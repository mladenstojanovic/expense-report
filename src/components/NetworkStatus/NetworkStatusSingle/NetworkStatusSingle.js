import React from 'react';
import { PropTypes } from 'prop-types';
import { SUCCESS } from '../../../store/actions/network/network.constants';
import loader from '../../../assets/icons/loader.svg';
import success from '../../../assets/icons/success.svg';
import { NetworkStatusSingleStyle } from '../NetworkStatus.style';

const NetworkStatusSingle = props => (
  <NetworkStatusSingleStyle>
    <img
      src={props.networkStatus !== SUCCESS ? loader : success}
      alt="loader"
    />{' '}
    <h2>{props.networkTitle}</h2>
  </NetworkStatusSingleStyle>
);

NetworkStatusSingle.propTypes = {
  networkStatus: PropTypes.string.isRequired,
  networkTitle: PropTypes.string.isRequired
};

export default NetworkStatusSingle;
