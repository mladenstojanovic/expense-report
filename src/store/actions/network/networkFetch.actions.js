import {
  networkOperationsStart,
  getAuthTokenStart,
  getAuthTokenError,
  createUserStart,
  createUserError,
  createUserSuccess,
  addBankConnectionStart,
  addBankConnectionError,
  addBankConnectionSuccess,
  connectionJobStart,
  connectionJobSuccess,
  connectionJobError,
  getTransactionsStart,
  getTransactionsError,
  getAuthTokenSuccess,
  getTransactionsSuccess
} from './network.actions';
import { sleep, sortTransactionData } from '../../../utils/utils';
import { RETRIEVE_TRANSACTIONS, SUCCESS } from './network.constants';

let userId;
let token;
let parsedAddConnectionResponse;
let jobFinished = false;

export const getAuthToken = async dispatch => {
  try {
    dispatch(networkOperationsStart());
    dispatch(getAuthTokenStart());

    const authResponse = await fetch(process.env.REACT_APP_TOKEN_ENDPOINT, {
      headers: {
        Authorization: `Basic ${process.env.REACT_APP_API_KEY}`
      }
    });

    const authResponseParsed = await authResponse.json();

    token = authResponseParsed.access_token;

    if (!token) {
      dispatch(getAuthTokenError());
      return {
        error: true
      };
    }

    dispatch(getAuthTokenSuccess());
    return {
      error: false
    };
  } catch {
    dispatch(getAuthTokenError());
    return {
      error: true
    };
  }
};

export const createUser = async (dispatch, userData) => {
  try {
    dispatch(createUserStart());

    const createUserResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users`,
      {
        method: 'POST',
        body: JSON.stringify(userData),
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const createUserResponseJson = await createUserResponse.json();

    userId = createUserResponseJson.id;

    if (!userId) {
      dispatch(createUserError());
      return {
        error: true
      };
    }
    dispatch(createUserSuccess());
    return {
      error: false
    };
  } catch {
    dispatch(createUserError());
    return {
      error: true
    };
  }
};

export const addConnection = async dispatch => {
  try {
    const testUser = {
      loginId: process.env.REACT_APP_LOGIN_ID,
      password: process.env.REACT_APP_LOGIN_PWD,
      institution: {
        id: process.env.REACT_APP_LOGIN_INSTITUTION_ID
      }
    };

    dispatch(addBankConnectionStart());

    const addConnectionResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/connections`,
      {
        method: 'POST',
        body: JSON.stringify(testUser),
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      }
    );
    parsedAddConnectionResponse = await addConnectionResponse.json();

    const jobId = parsedAddConnectionResponse.id;

    if (!jobId) {
      dispatch(addBankConnectionError());
      return {
        error: true
      };
    }

    dispatch(addBankConnectionSuccess());
    return { error: false, parsedAddConnectionResponse };
  } catch {
    dispatch(addBankConnectionError());
    return {
      error: true
    };
  }
};

export const connectionJob = async (
  dispatch,
  pollingRetries,
  pollingTime,
  parsedAddConnectionResponse
) => {
  try {
    dispatch(connectionJobStart());
    let retries = pollingRetries;
    do {
      await sleep(pollingTime);

      const job = await fetch(`${parsedAddConnectionResponse.links.self}`, {
        headers: {
          Authorization: `Bearer ${token}`,
          'content-type': 'application/json'
        }
      });
      const parsedJob = await job.json();
      const retrieveTransactionJob = parsedJob.steps.find(
        job => job.title === RETRIEVE_TRANSACTIONS
      );

      if (retrieveTransactionJob.status === SUCCESS) {
        jobFinished = true;
      } else {
        retries--;
      }
    } while (!jobFinished && retries > 0);
    if (jobFinished) {
      dispatch(connectionJobSuccess());
      return true;
    } else {
      dispatch(connectionJobError());
      return false;
    }
  } catch {
    dispatch(connectionJobError());
    return false;
  }
};

export const getTransactions = async dispatch => {
  try {
    dispatch(getTransactionsStart());
    const getTransactionsResponse = await fetch(
      `${process.env.REACT_APP_BASE_ENDPOINT}/users/${userId}/transactions`,
      {
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    const parsedGetTransactionsResponse = await getTransactionsResponse.json();

    if (!parsedGetTransactionsResponse.data) {
      dispatch(getTransactionsError());
      return;
    }

    dispatch(
      getTransactionsSuccess(
        sortTransactionData(parsedGetTransactionsResponse.data)
      )
    );
  } catch {
    dispatch(getTransactionsError());
  }
};
