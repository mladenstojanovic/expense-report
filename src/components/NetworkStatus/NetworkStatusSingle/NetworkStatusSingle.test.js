import React from 'react';
import { renderWithReduxAndStyles } from '../../../utils/utils';
import NetworkStatusSingle from './NetworkStatusSingle';
import { SUCCESS } from '../../../store/actions/network/network.constants';

describe('Test NetworkStatusSingle component', () => {
  it('should render correctly', () => {
    const networkTitleText = 'test network title text';
    const { container, getByText, getByAltText } = renderWithReduxAndStyles(
      <NetworkStatusSingle
        networkStatus="test"
        networkTitle={networkTitleText}
      />
    );

    const loader = getByAltText('loader');

    expect(loader.src).toContain('loader.svg');
    expect(getByText(networkTitleText)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly when network status is success', () => {
    const networkTitleText = 'test network title text';
    const { container, getByText, getByAltText } = renderWithReduxAndStyles(
      <NetworkStatusSingle
        networkStatus={SUCCESS}
        networkTitle={networkTitleText}
      />
    );

    const loader = getByAltText('loader');

    expect(loader.src).toContain('success.svg');
    expect(getByText(networkTitleText)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
