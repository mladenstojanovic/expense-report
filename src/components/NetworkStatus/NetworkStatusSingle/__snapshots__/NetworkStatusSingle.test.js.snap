// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NetworkStatusSingle component should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <img
    alt="loader"
    src="loader.svg"
  />
   
  <h2>
    test network title text
  </h2>
</div>
`;

exports[`Test NetworkStatusSingle component should render correctly when network status is success 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <img
    alt="loader"
    src="success.svg"
  />
   
  <h2>
    test network title text
  </h2>
</div>
`;
